### ISSUE TITLE
Random NaN loss when using float16 dtype and batch size of 1

### ISSUE THREAD

**Issue by PaulPauls** [tensorflow/tensorflow#38457]
__System information__ 
- Platform: Ubuntu Linux (Kernel 5.3) with Python 3.6.9 **OR** Google Colab
- Tested on TensorFlow: TF v2.1.0 and TF v2.2.0rc2

__Background__
I came across the following bug in one of my Tensorflow projects and was able to successfully reproduce the bug with minimal code in a Google colab. Please see the link to this colab below. The execution of this colab also shows the bug occuring in iteration 248.

__Reproducing Code__
see: [https://colab.research.google.com/drive/1ZzeqGSOKL5qw9j7XPqdlkHQvFAZETU5O](https://colab.research.google.com/drive/1ZzeqGSOKL5qw9j7XPqdlkHQvFAZETU5O)

```
import math
import numpy as np
import tensorflow as tf


if __name__ == '__main__':

    x = np.array([[0, 0], [0, 1], [1, 0], [1, 1]])
    y = np.array([[0], [1], [1], [0]])

    loss_function = tf.keras.losses.BinaryCrossentropy()

    for i in range(2000):
        if i % 100 == 0:
            print("Iteration {}".format(i))

        model = tf.keras.Sequential([
            tf.keras.layers.Dense(units=1, activation='tanh', dtype=tf.float16)])

        model.compile(optimizer='sgd', loss=loss_function)
        model.fit(x=x, y=y, epochs=1, batch_size=1)

        loss_result = loss_function(y, model(x))

        if math.isnan(loss_result):
            raise RuntimeError('NAN Error in iteration {}'.format(i))
```

__Behaviour Description__
Seemingly non-deterministic occurence of a NaN result when calculating loss of a very simple Dense Model. The NaN loss seems to happen randomly and can occur on the 60th or 600th iteration. In the supplied Google colab code it happened in the 248th iteration. The bug only seems to occur using a dtype of float16 and batch_size of 1. Debugging the error lead me to see that the models producing a NaN loss seem to have been initialized with a NaN bias and kernel, though I couldn't get to the bottom of why.

**Comment by Saduf2019** [comment]
i am able to replicate [this issue](https://colab.sandbox.google.com/gist/Saduf2019/145f094f4cab5ef14028a050aa2c53e3/38461.ipynb)

**Comment by jvishnuvardhan** [comment]
@PaulPauls I think we need to use [`mixed_precision`](https://www.tensorflow.org/guide/mixed_precision) to avoid any instability issues coming due to mixed precision. 

When I added the following line to your code, it doesn't throw the error you are noticing.
```
from tensorflow.keras.mixed_precision import experimental as mixed_precision

policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)
```

Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/3b61ebeab8c397349b5c22c2e83f5fe6/38461.ipynb) with CPU and [here](https://colab.research.google.com/gist/jvishnuvardhan/b445ffe0bccf6a40bf091cb41694402c/untitled27.ipynb) is with GPU. Thanks!

**Comment by PaulPauls** [comment]
Thanks for offering a hotfix @jvishnuvardhan! I tried out setting the ``mixed_precision`` policy and it did indeed solve the problem, however the performance loss is considerable. 

I tried out timing the [original code](https://colab.research.google.com/drive/19hryWYHCGhDvzNh3DwtGLPFESuMmZ0Tt?usp=sharing) vs the [hotfixed code](https://colab.research.google.com/drive/1nH4WymavPJKVWTK56oU4_dpfxlXxysgI?usp=sharing) and the hotfixed code required nearly twice as much processing time on CPU (152s vs 260s). Similar results when using a GPU: I couldn't get access to a colab GPU as they were all being used but when I timed in on my private machine with a GeForce 2070 the timing results were 92s vs 147s.

Effectively for me this means that it is more sensible to check for the occasional NaN result and recreate the model rather than permanently change the mixed precision policy. 

Still, thanks for taking on the issue! I hope others will benefit from this hotfix. =)

**Comment by jvishnuvardhan** [comment]
@PaulPauls We cannot compare the performance with CPU as the `mixed_precision API` throws a warning as follows

> WARNING:tensorflow:Mixed precision compatibility check (mixed_float16): WARNING
> The dtype policy mixed_float16 may run slowly because this machine does not have a GPU. Only Nvidia GPUs with compute capability of at least 7.0 run quickly with mixed_float16.
> If you will use compatible GPU(s) not attached to this host, e.g. by running a multi-worker model, you can ignore this warning. This message will only be logged once

Regarding GPU performance with `mixed_precision API`, may be you can open another issue.

I am closing this issue as the original issue was resolved. Please feel free to reopen if I am mistaken. thanks!
